File: c/array/binary_search.c
#include <stdio.h>
#include <stdlib.h>

/*
 * [left, right]
 * 
 * */

int binary_search(int nums[], int arrlen, int target);

// gcc binary_search.c -o bs
// ./bs <target>
int main(int argc, char *argv[]) {
  int nums[] = {1, 3, 12, 44, 65, 213};

  int arrlen = sizeof(nums) / sizeof(nums[0]);
  int target = atoi(argv[1]);

  printf("%d\n", binary_search(nums, arrlen, target));
  return 0;
}

int binary_search(int nums[], int arrlen, int target) {
  int low, high, middle;

  low = 0;
  high = arrlen;
  middle = (low + high) / 2;

  while(low <= high) {

    if (nums[middle] < target) {
      low = middle+1;
    } else if (nums[middle] > target){
      high = middle-1;
    } else {
      return nums[middle];
    }

    middle = (low + high) / 2;
  }

  return -1;
}

File: python/Binary_search.py
# 该函数返回的是 元素 在列表中的索引

def binary_search(list, item):

    # low and high 代表索引
    low = 0
    high = len(list)-1

    while low <= high:
        mid = int((low+high)/2)
        guess = list[mid]

        if guess == item:
            return mid

        if guess < item:

            # +1 是把 mid 排除出去 因为已经知道 mid 不在范围内
            low = mid + 1
        else:

            # -1 也是为了把 mid 排除出去
            high = mid - 1

    return None

my_list = [1, 3, 5, 7, 9,10,11,23]

print (binary_search(my_list, 11))
